app/services/solutions.rb

```diff
@@ -0,0 +1,100 @@
+require_relative '../repositories/repository_challenge'
+require_relative './solution'
+
+class Solutions
+
+  PER_PAGE = 100
+
+  def initialize(player, challenge_id, page)
+    @player = player
+    @challenge_id = challenge_id
+    @page = (page || 1).to_i
+  end
+  attr_reader :player
+  attr_reader :challenge_id
+  attr_reader :page
+
+  def each
+    position = PER_PAGE * (page-1) + count_remaining + 1
+
+    solutions.each_with_index do |s, i|
+      yield Solution.new(s, users, position + i)
+    end
+  end
+
+  def users
+    @users ||= User.where(:_id.in => user_ids).inject({}) {|h,u| h.merge(u.id => u)}
+  end
+
+  def user_ids
+    solutions.map {|c| c[:user_id] }
+  end
+
+  def solutions
+    @solutions ||= RepositoryChallenge.solutions(
+      challenge_id: challenge_id,
+      min_score: visible_score,
+      per_page: PER_PAGE,
+      page: page
+    )
+  end
+
+  def visible_score
+    @visible_score ||= if !player
```

Comment:
Could we unroll this one into multi-line? Easier to grok :)

Reply:
done